### PR TITLE
enhancement(remap): introduce `is_TYPE` helper functions

### LINKED ISSUE #6732: Add `is_TYPE` functions to vrl
Noted over here, https://github.com/timberio/vector/pull/6709#discussion_r592616329, that it is difficult to check the type of a value right now. You can use the coercion functions and check for an error, but this is a bit awkward:

```
_, err = object(.foo)
foo_is_object = (err == nil)
```

I suggest we add functions like `is_object` which would just return a boolean.

```
foo_is_object = is_object(.foo)
```

### PR DESCRIPTION
**Pull request by prognant**
Closes #6732

Introduce a set of helper functions to test the type of a VRL var:

`is_string("foobar")` returns `true` 

`is_string(1515)` returns `false` 

`is_integer(11)` returns `true`  

and so on for the supported types. 

Signed-off-by: prognant <pierre.rognant@datadoghq.com>

### CODE REVIEW

File: lib/vrl/stdlib/src/type_of.rs [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+            Value::Null => Ok(Value::Bytes(bytes::Bytes::from("null"))),
+        }
+    }
+
+    fn type_def(&self, _: &state::Compiler) -> TypeDef {
+        TypeDef::new().infallible().bytes()
+    }
+}
+
+/*#[cfg(test)]
```
Comment (jszwedko) [inline lib/vrl/stdlib/src/type_of.rs:60]: I think the unit tests should be runnable now as-of #6492

**Review by JeanMertz — COMMENTED** [review]
I haven't looked at the implementation details yet, but before I do that, I wanted to note that a previous issue was filed to discuss this feature:

https://github.com/timberio/vector/issues/5261

In it, @binarylogic also [proposed](https://github.com/timberio/vector/issues/5261#issuecomment-754166749) a `type` function similar to `type_of`.

I raised a concern that this could become problematic in the future if we want to introduce type-level scoping.

For example, I envision us being able to do:

```rust
if is_string(.foo) {
	// compiler knows `.foo` is a string, so this is OK
	.foo = upcase(.foo)
}
```

For this to work, we'd need to add a feature to the compiler that allows functions to bubble up type information invariants about one or more of their arguments. In this case, the `is_string` function would let the compiler know whether `.foo` is a string, which can then be used within the `if block` scope to allow functions to treat `.foo` as if it were a string.

By having a `type_of` function, this becomes more difficult and/or hacky to solve because it would no longer _just_ be the function that can inform the compiler about the type, you'd also need to special-case the conditional that happens afterward (e.g. `type_of(.foo) == "string")`.

So, my preference would be to add all missing `is_*` functions, then more forward with the type-level scoping proposal, and after that make a proposal on how we could add `type_of`.

**Comment by prognant** [comment]
@JeanMertz Thanks for the insights ! I read the discussion in #5261 and based on my somehow very low knowledge of VRL I concur with the conclusion that `is_<type>` functions are a better option as of today. And ultimately after having a type-level scoping, implement a proper `type_of` function that could properly type-feed a conditional block.

I will rework the PR accordingly.

File: lib/vrl/stdlib/src/is_array.rs [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    value: Box<dyn Expression>,
+}
+
+impl Expression for IsArrayFn {
+    fn resolve(&self, ctx: &mut Context) -> Resolved {
+        let v = self.value.resolve(ctx)?;
+        match v {
+            Value::Array(_) => Ok(value!(true)),
+            _ => Ok(value!(false)),
+        }
```
Comment (FungusHumungus) [inline lib/vrl/stdlib/src/is_array.rs:57]: This code is fine, but could be simplified using [matches!](https://doc.rust-lang.org/std/macro.matches.html) to something like:

```rust
Ok(value!(matches!(v, Value::Array(_))))
```

**Comment by FungusHumungus** [comment]
I think it would be worth adding an `is_null` function as well. Currently we have `is_nullish`, but that does a bit more than a simple `is_null` would do. eg.. `is_nullish("") == true` but `is_null("") == false`.

**Comment by prognant** [comment]
> I think it would be worth adding an `is_null` function as well. Currently we have `is_nullish`, but that does a bit more than a simple `is_null` would do. eg.. `is_nullish("") == true` but `is_null("") == false`.

Thanks for the review !
I was thinking that `some_val == null` could be used for null type assertion (but I'm not 100% sure it would work) anyway for consistency purpose I agreee that having an `is_null` function make sense. I will add it.

**Comment by FungusHumungus** [comment]
> I was thinking that `some_val == null`

Yes, of course, I hadn't thought of that. 

I think it might still come in handy for future work as Jean suggested earlier where we may then be able to derive typing information inside the if blocks.

File: docs/reference/remap/functions/is_timestamp.cue [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+			description: #"The value to check"#
+			required:    true
+			type: ["any"]
+		},
+	]
+	internal_failure_reasons: []
+	return: {
+		types: ["boolean"]
+		rules: [
+			#"Returns `true` if `value` is a ."#,
```
Comment (JeanMertz) [inline docs/reference/remap/functions/is_timestamp.cue:21]:
```suggestion
			#"Returns `true` if `value` is a timestamp."#,
```

File: lib/vrl/stdlib/src/is_array.rs (hunk 2 above)
Comment (JeanMertz) [reply lib/vrl/stdlib/src/is_array.rs:57]: At that point, you might as well collapse it all together 😄 

`Value` has `is_xxx()` already implemented, so you can do:

```rust
self.value.resolve(ctx).map(Value::is_array)
```

**Review by JeanMertz — APPROVED** [review]
LGTM! I added one comment, but I don't consider that a blocker.

**Review by jszwedko — COMMENTED** [review]
This looks good to me! Could we add these to the benchmark suite too?

https://github.com/timberio/vector/blob/07a5427677934be4270fe39de31ec5b460344722/lib/vrl/stdlib/benches/benches.rs

**Comment by prognant** [comment]
Sure will do.

**Review by jszwedko — APPROVED** [review]
🎉

File: docs/reference/remap/functions/is_null.cue [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+			required:    true
+			type: ["any"]
+		},
+	]
+	internal_failure_reasons: []
+	return: {
+		types: ["boolean"]
+		rules: [
+			#"Returns `true` if `value` is null."#,
+			#"Returns `false` if `value` is anything else."#,
```
Comment (pablosichert) [inline docs/reference/remap/functions/is_null.cue:23]:
```suggestion
			"Returns `true` if `value` is null.",
			"Returns `false` if `value` is anything else.",
```

File: docs/reference/remap/functions/is_object.cue [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+			required:    true
+			type: ["any"]
+		},
+	]
+	internal_failure_reasons: []
+	return: {
+		types: ["boolean"]
+		rules: [
+			#"Returns `true` if `value` is an object."#,
+			#"Returns `false` if `value` is anything else."#,
```
Comment (pablosichert) [inline docs/reference/remap/functions/is_object.cue:22]:
```suggestion
			"Returns `true` if `value` is an object.",
			"Returns `false` if `value` is anything else.",
```

File: docs/reference/remap/functions/is_regex.cue [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+remap: functions: is_regex: {
+	category: "Type"
+	description: """
+		Check if the type of a `value` is a regex or not.
+		"""
+
+	arguments: [
+		{
+			name:        "value"
+			description: #"The value to check"#
```
Comment (pablosichert) [inline docs/reference/remap/functions/is_regex.cue:12]:
```suggestion
			description: "The value to check."
```

File: docs/reference/remap/functions/is_regex.cue [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+			required:    true
+			type: ["any"]
+		},
+	]
+	internal_failure_reasons: []
+	return: {
+		types: ["boolean"]
+		rules: [
+			#"Returns `true` if `value` is a regex."#,
+			#"Returns `false` if `value` is anything else."#,
```
Comment (pablosichert) [inline docs/reference/remap/functions/is_regex.cue:22]:
```suggestion
			"Returns `true` if `value` is a regex.",
			"Returns `false` if `value` is anything else.",
```

File: docs/reference/remap/functions/is_string.cue [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+remap: functions: is_string: {
+	category: "Type"
+	description: """
+		Check if the type of a `value` is a string or not.
+		"""
+
+	arguments: [
+		{
+			name:        "value"
+			description: #"The value to check"#
```
Comment (pablosichert) [inline docs/reference/remap/functions/is_string.cue:12]:
```suggestion
			description: "The value to check."
```

File: docs/reference/remap/functions/is_string.cue [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+			required:    true
+			type: ["any"]
+		},
+	]
+	internal_failure_reasons: []
+	return: {
+		types: ["boolean"]
+		rules: [
+			#"Returns `true` if `value` is a string."#,
+			#"Returns `false` if `value` is anything else."#,
```
Comment (pablosichert) [inline docs/reference/remap/functions/is_string.cue:22]:
```suggestion
			"Returns `true` if `value` is a string.",
			"Returns `false` if `value` is anything else.",
```

File: docs/reference/remap/functions/is_timestamp.cue [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+remap: functions: is_timestamp: {
+	category: "Type"
+	description: """
+		Check if the type of a `value` is a timestamp or not.
+		"""
+
+	arguments: [
+		{
+			name:        "value"
+			description: #"The value to check"#
```
Comment (pablosichert) [inline docs/reference/remap/functions/is_timestamp.cue:12]:
```suggestion
			description: "The value to check."
```

File: docs/reference/remap/functions/is_timestamp.cue [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+			required:    true
+			type: ["any"]
+		},
+	]
+	internal_failure_reasons: []
+	return: {
+		types: ["boolean"]
+		rules: [
+			#"Returns `true` if `value` is a timestamp."#,
+			#"Returns `false` if `value` is anything else."#,
```
Comment (pablosichert) [inline docs/reference/remap/functions/is_timestamp.cue:22]:
```suggestion
			"Returns `true` if `value` is a timestamp.",
			"Returns `false` if `value` is anything else.",
```

**Review by pablosichert — COMMENTED** [review]
Just some pedantic suggestions for the docs 😄

**Pull request merged**